packages/react/src/Banner/Banner.tsx

```diff
@@ -0,0 +1,323 @@
+import cx from 'clsx'
+import React, {createContext, useContext, useEffect, useId, useMemo} from 'react'
+import styled from 'styled-components'
+import {AlertIcon, InfoIcon, StopIcon, CheckCircleIcon, XIcon} from '@primer/octicons-react'
+import {Button, IconButton} from '../Button'
+import {get} from '../constants'
+
+type BannerVariant = 'critical' | 'info' | 'success' | 'upsell' | 'warning'
+
+export type BannerProps = React.ComponentPropsWithoutRef<'section'> & {
+  /**
+   * Provide an optional description for the Banner. This should provide
+   * supplemental information about the Banner
+   */
+  description?: React.ReactNode
+
+  /**
+   * Provide an icon for the banner.
+   * Note: Only `variant="info"` banners should use custom icons
+   */
+  icon?: React.ReactNode
+
+  /**
+   * Optionally provide a handler to be called when the banner is dismissed.
+   * Providing this prop will show a dismiss button.
+   *
+   * Note: This is not available for critical banners.
+   */
+  onDismiss?: () => void
+
+  /**
+   * Provide an optional primary action for the Banner.
+   */
+  primaryAction?: React.ReactNode
+
+  /**
+   * Provide an optional secondary action for the Banner
+   */
+  secondaryAction?: React.ReactNode
+
+  /**
+   * The title for the Banner. This will be used as the accessible name and is
+   * required unless `Banner.Title` is used as a child.
+   */
+  title?: React.ReactNode
+
+  /**
+   * Specify the type of the Banner
+   */
+  variant?: BannerVariant
+}
+
+const iconForVariant: Record<BannerVariant, React.ReactNode> = {
+  critical: <StopIcon />,
+  info: <InfoIcon />,
+  success: <CheckCircleIcon />,
+  upsell: <InfoIcon />,
+  warning: <AlertIcon />,
+}
+
+export const Banner = React.forwardRef<HTMLElement, BannerProps>(function Banner(
+  {children, description, icon, onDismiss, primaryAction, secondaryAction, title, variant = 'info', ...rest},
+  ref,
+) {
+  const titleId = useId()
+  const value = useMemo(() => {
+    return {
+      titleId,
+    }
+  }, [titleId])
+  const dismissible = variant !== 'critical' && onDismiss
+  const hasActions = primaryAction || secondaryAction
+
+  if (__DEV__) {
+    // Note: __DEV__ will make it so that this hook is consistently called, or
+    // not called, depending on environment
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    useEffect(() => {
+      const title = document.getElementById(titleId)
```

Comment:
To circle back on - https://github.com/github/primer/pull/3090#discussion_r1523866786!